server/engine/chainer_test.go

```diff
@@ -0,0 +1,884 @@
+			)
+			Convey("When I try to remove an empty workflow", func() {
+				err := chainer.Remove("")
+				Convey("Then I get an error telling the workflow is empty", func() {
+					So(err, ShouldNotBeNil)
+					So(err.Error(), ShouldContainSubstring, "is empty")
+				})
+			})
+			Convey("When I try to remove a workflow that does not exists", func() {
+				err := chainer.Remove("not_existing")
+				Convey("Then the existing workflow is still here", func() {
+					So(err, ShouldBeNil)
+					So(chainer.workflows, ShouldHaveLength, 1)
+				})
+			})
+			Convey("When I try to remove an existing workflow", func() {
+				err := chainer.Remove("test")
+				Convey("Then the chainer should not contain any workflow anymore", func() {
+					So(err, ShouldBeNil)
+					So(chainer.workflows, ShouldHaveLength, 0)
+				})
+			})
+		})
+	})
+}
+
+func TestChainerEngineSimpleOKWorkflow(t *testing.T) {
+	Convey("On a docktor chainer engine", t, func() {
+		Convey("Given a chainer engine with a simple workflow", func() {
+			chainer := NewChainEngine()
+			context := &ChainerContext{
+				Data: map[string]interface{}{
+					"test.up":   []string{},
+					"test.down": []string{},
+				},
+			}
+			chainer.Add(
+				"test",
+				Step{Up: StepUpOK("test"), Down: StepDownOK("test")},
+				Step{Up: StepUpOK("test"), Down: StepDownOK("test")},
+				Step{Up: StepUpOK("test"), Down: StepDownOK("test")},
+				Step{Up: StepUpOK("test"), Down: StepDownOK("test")},
+			)
+			up := Operation{
+				Errors: make(chan error),
+				Status: make(chan string),
+			}
+			down := Operation{
+				Errors: make(chan error),
+				Status: make(chan string),
+			}
+
+			done := make(chan bool)
+
+			Convey("When I run a workflow that does not exist", func() {
+				go chainer.Run("dontexist", context, up, down, done)
+
+				var nbStepOK, nbStepKO int
+				Convey("Then I get errors saying that workflow does not exist", func() {
+					run := true
+					for run {
+						select {
+						case <-done:
+							run = false
+						case _, more := <-up.Status:
+							if more {
+								nbStepOK++
+							}
+						case err, more := <-up.Errors:
+							if more {
+								So(err.Error(), ShouldContainSubstring, "does not exist")
+								nbStepKO++
+							}
+						case _, more := <-down.Status:
+							if more {
+								nbStepOK++
+							}
+						case err, more := <-down.Errors:
+							if more {
+								So(err.Error(), ShouldContainSubstring, "does not exist")
+								nbStepKO++
+							}
+						}
+					}
+					Convey("and a single step has been executed and is KO", func() {
+						So(nbStepOK, ShouldEqual, 0)
+						So(nbStepKO, ShouldEqual, 1)
+					})
+				})
+			})
+
+			Convey("But When I run an existing workflow, and try to add a new one while chain is executing", func() {
+				go chainer.Run("test", context, up, down, done)
+
+				var nbStepOK, nbStepDownAndErrors int
+				Convey("Then I get errors saying that I can't add a workflow while running", func() {
+					run := true
```

Comment:
non mais làtu devrais pas avoir de traitement differents que des resultats differents ou allors t'as foiré ton engine :).
Mais je suis d'accord les méthodes génériques c'est pas forcement nécéssaire mais pour des tests c'est la lisibilité qui prime si t'as du traitement degueulasse faut l'encapsuler.

func handleWorkflow(up Operation, down Operation) (nbUpOK int, nbUpKO int, nbDownOK int, nbDownKO int, errs []error) {
	nbUpOK = 0
	nbUpKO = 0
	nbDownOK = 0
	nbDownKO = 0
	errs = []errors{}
	run := true
	for run {
		select {
		case <-done:
			run = false
		case _, more := <-up.Status:
			if more {
				nbUpOK++
			}
		case err, more := <-up.Errors:
			if more {
				errs = append(errs, err)
				nbUpKO++
			}
		case _, more := <-down.Status:
			if more {
				nbDownOK++
			}
		case err, more := <-down.Errors:
			if more {
				errs = append(errs, err)
				nbDownKO++
			}
		}
	}
	return
}
et ensuite tu fais tes So